web/src/main/java/org/springframework/security/web/csrf/SecureCsrfToken.java

```diff
@@ -0,0 +1,151 @@
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.web.csrf;
+
+import java.nio.ByteBuffer;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.util.Base64;
+import org.springframework.security.crypto.codec.Utf8;
+import org.springframework.util.Assert;
+
+/**
+ * A CSRF token that is used to protect against CSRF attacks with extra protection from BREACH exploit
+ *
+ * @author Ruby Hartono
+ * @since 5.4
+ */
+@SuppressWarnings("serial")
+public final class SecureCsrfToken implements CsrfToken {
+
+	private final byte[] tokenBytes;
+
+	private final String parameterName;
+
+	private final String headerName;
+
+	private final SecureRandom secureRandom;
+
+	private final int randomByteSize;
+
+	/**
+	 * Creates a new instance
+	 *
+	 * @param headerName    the HTTP header name to use
+	 * @param parameterName the HTTP parameter name to use
+	 * @param token         the value of the token (i.e. expected value of the HTTP
+	 *                      parameter of parametername).
+	 */
+	public SecureCsrfToken(String headerName, String parameterName, String token) {
+		Assert.hasLength(headerName, "headerName cannot be null or empty");
+		Assert.hasLength(parameterName, "parameterName cannot be null or empty");
+		Assert.hasLength(token, "token cannot be null or empty");
+		this.headerName = headerName;
+		this.parameterName = parameterName;
+		this.tokenBytes = Utf8.encode(token);
+		this.secureRandom = new SecureRandom();
+		this.randomByteSize = this.tokenBytes.length;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getHeaderName()
+	 */
+	public String getHeaderName() {
+		return this.headerName;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getParameterName()
+	 */
+	public String getParameterName() {
+		return this.parameterName;
+	}
+
+	/*
+	 * (non-Javadoc)
+	 *
+	 * @see org.springframework.security.web.csrf.CsrfToken#getToken()
+	 */
+	public String getToken() {
+		ByteBuffer byteBuffer = ByteBuffer.allocate(this.randomByteSize);
+		this.secureRandom.ints(Byte.MIN_VALUE, Byte.MAX_VALUE).limit(randomByteSize)
+				.forEach((randInt) -> byteBuffer.put((byte) randInt));
+
+		byte[] randomBytes = byteBuffer.array();
+
+		byte[] xoredCsrf = xorCsrf(randomBytes, this.tokenBytes);
+
+		ByteBuffer combinedBuffer = ByteBuffer.allocate(this.randomByteSize + xoredCsrf.length);
+		combinedBuffer.put(randomBytes);
+		combinedBuffer.put(xoredCsrf);
+
+		// returning randomBytes + XOR csrf token
+		return Base64.getEncoder().encodeToString(combinedBuffer.array());
```

Comment:
I'd prefer to avoid the use of `ByteBuffer` if possible. This will avoid unnecessary overhead (object creation which will mean more Garbage Collection) in a method that is used heavily within the system. It seems that `System.arraycopy` would work well.

```suggestion
		byte[] combinedBytes = new byte[randomBytes.length + xoredCsrf.length];
		System.arraycopy(randomBytes, 0, combinedBytes, 0, randomBytes.length);
		System.arraycopy(xoredCsrf, 0, combinedBytes, randomBytes.length, xoredCsrf.length);

		// returning randomBytes + XOR csrf token
		return Base64.getEncoder().encodeToString(combinedBytes);